Update application howtos [WD-19134] (#288)
canonical/anbox-cloud-docs: howto/application/update-application.md

  disk-size: 8GB
```

````

```{group-tab} Dashboard

> To update an application with Anbox Cloud dashboard, visit the *Applications* page and find the application you want to edit in the table. Click the *Edit* button (🖉) in the actions column to open the *Update application* page.

Author: Or maybe like this:
```suggestion
You can edit an application from the *Applications* list page, or from an application's detail page.
```
So that we can incorporate also @anusha-c18's suggestion?

> To update an application with Anbox Cloud dashboard, visit the *Applications* page and find the application you want to edit in the table. Click the *Edit* button (🖉) in the actions column to open the *Update application* page.

Reviewer: @lorumic The updates have added value overall. For the remaining two outstanding comments (including this one), I am fine with Anusha's suggestion. Once you are done applying them, we can approve and merge.

> To update an application with Anbox Cloud dashboard, visit the *Applications* page and find the application you want to edit in the table. Click the *Edit* button (🖉) in the actions column to open the *Update application* page.

Author: I'm a bit confused, from [this comment](https://github.com/canonical/anbox-cloud-docs/pull/288#discussion_r1954017341) I thought that I should have followed your suggestion above, and just include one way to reach the page?

> To update an application with Anbox Cloud dashboard, visit the *Applications* page and find the application you want to edit in the table. Click the *Edit* button (🖉) in the actions column to open the *Update application* page.

Author: Ah, I see, and noticed your edit just now. Ok, let's wait for @anusha-c18 as I have no strong opinion on this. I made the proposal above to try to get the best of both your suggestions :)

> To update an application with Anbox Cloud dashboard, visit the *Applications* page and find the application you want to edit in the table. Click the *Edit* button (🖉) in the actions column to open the *Update application* page.

Reviewer: @anusha-c18 If you are happy with this, we can merge

> To update an application with Anbox Cloud dashboard, visit the *Applications* page and find the application you want to edit in the table. Click on the pencil icon (🖉) in the actions column, and this will bring you to the *Update application* page.

Reviewer 2: We can also include a line on editing the application from the application details page

> To update an application with Anbox Cloud dashboard, visit the *Applications* page and find the application you want to edit in the table. Click on the pencil icon (🖉) in the actions column, and this will bring you to the *Update application* page.

Author: Better using "open" instead of "view". And I don't think there is a need to include all the possible ways to reach the page. One is enough.

> To update an application with Anbox Cloud dashboard, visit the *Applications* page and find the application you want to edit in the table. Click on the pencil icon (🖉) in the actions column, and this will bring you to the *Update application* page.

Reviewer 2: > Better using "open" instead of "view". 

Sure.

> And I don't think there is a need to include all the possible ways to reach the page. One is enough.

We do list it in the images page's documentation. 
```
To update an application, click the *Edit* button () on the *Applications* page or the *Application details* page.
```

> To update an application with Anbox Cloud dashboard, visit the *Applications* page and find the application you want to edit in the table. Click on the pencil icon (🖉) in the actions column, and this will bring you to the *Update application* page.

Reviewer: Introducing options of doing the same thing in multiple ways is fine but if we do that in every instance, it gets too redundant. I'd suggest following the comment later.

> To update an application with Anbox Cloud dashboard, visit the *Applications* page and find the application you want to edit in the table. Click on the pencil icon (🖉) in the actions column, and this will bring you to the *Update application* page.

Author: Ok, so I'll stick to your original suggestion there.

> In the guided form, you can customize almost all the fields that can be customized during {ref}`application creation <howto-create-application>`. Some of the fields, like the application name, or the container/vm option, are not meant to be customized, so they are disabled. When fields are disabled, a tooltip that appears hovering over them will explain the reason.

Reviewer 2: ```suggestion
In the guided form, you can customize almost all the fields that can be customized during {ref}`application creation <howto-create-application>`. Some of the fields, like the application name, or the container/VM option, are not meant to be customized, so they are disabled. When fields are disabled, a tooltip that appears hovering over them will explain the reason.
```

> Just like in the creation case, you can unlock advanced customization capabilities by toggling the *Customize manifest.yaml* switch at the bottom of the page. It is important to mention that both manifest.yaml customization, upload of a new APK as well as the customization of some fields under *Configuration (optional)* ({ref}`manifest version name <ref-application-manifest>`, {ref}`boot activity <ref-application-manifest>`, {ref}`features <ref-feature-flags>` and {ref}`sec-application-manifest-watchdog`) will trigger the creation of a new application version.

Author: I have rephrased this so that there is no need to list all the fields. Instead, it should not be omitted that uploading an APK and customizing the manifest.yaml will also trigger the creation of a new application version.

> Just like in the creation case, you can unlock advanced customization capabilities by toggling the *Customize manifest.yaml* switch at the bottom of the page. It is important to mention that both manifest.yaml customization, upload of a new APK as well as the customization of some fields under *Configuration (optional)* ({ref}`manifest version name <ref-application-manifest>`, {ref}`boot activity <ref-application-manifest>`, {ref}`features <ref-feature-flags>` and {ref}`sec-application-manifest-watchdog`) will trigger the creation of a new application version.

Reviewer 2: I prefer listing the specific fields that trigger the creation of a new version, as it allows users to get all the necessary information directly from this section.

> Once finished updating the fields, you can submit the request by clicking on the *Update* button at the bottom of the page. You will be redirected either to the *Overview* tab of the application detail page - if the fields changed did not trigger a new version creation - or to the *Versions* tab of that page, when a new version of the application was created.

Author: Well, that's maybe because this is the only case where you have a different path for the 2 different outcomes of the application update?

  disk-size: 8GB
```

````

```{group-tab} Dashboard

You can edit an application from the *Applications* list page.

In the guided form, you can customize most of the fields that are available during {ref}`application creation <howto-create-application>`. Some of the fields, like the application name, or the container/VM option, cannot be modified and are disabled. When a field is disabled, hovering over it will display a tooltip explaining the reason.

Similar to the application creation process, you can unlock advanced customization capabilities by toggling the *Customize manifest.yaml* switch at the bottom of the page. Note that customizing the manifest or updating one of the following fields will create a new version of the application:
* APK
* {ref}`manifest version name <ref-application-manifest>`
* {ref}`boot activity <ref-application-manifest>`
* {ref}`features <ref-feature-flags>` 
* {ref}`sec-application-manifest-watchdog`

> 

Reviewer 3: hmm, this is not the case when I check the backend code. 
https://github.com/canonical/ams/blob/main/internal/ams/backend/application.go#L639-L659

Reviewer 3: Sure, when I read the following [code](https://github.com/canonical/ams/blob/main/internal/ams/backend/application.go#L639-L659)
if one of the following fields is given during the application update, there would be a new app version to be created.
- Services
- Watchdog
- BootActivity
- RequiredPermissions
- VideoEncoder
- ManifestVersion
- Features
- Hooks
- Bootstrap

So we may need to extend the list a bit in the doc as currently we only mentioned 4 of them + APK

Author: This is about the dashboard, not the CLI nor the AMS API. The only fields that you can customise in the dashboard form are those listed there. The other fields are not included in the form.

Reviewer 3: Okay, thanks for the correction.
